File: 1938-maximum-genetic-difference-query/1938-maximum-genetic-difference-query.js
/**
 * @param {number[]} parents
 * @param {number[][]} queries
 * @return {number[]}
 */
var maxGeneticDifference = function(parents, queries) {
  const [m, n] = [parents.length, queries.length];
  const graph = Array.from({ length: m }, () => []);
  const queriesByNode = Array.from({ length: m }, () => []);
  const trie = new Trie();
  const ans = new Array(n);
  let root = -1;
  
  parents.forEach((parent, i) => {
    if (parent === -1) root = i;
    else graph[parent].push(i);
  })
  
  queries.forEach(([node, val], i) => queriesByNode[node].push([val, i]));
  
  dfs(root, trie, graph, queriesByNode, ans);
  
  return ans;
};

const dfs = (curr, trie, graph, queriesByNode, ans) => {
  trie.insert(curr, 1);
  queriesByNode[curr].forEach(([val, i]) => ans[i] = trie.findMaxXOR(val));
  graph[curr].forEach((next) => dfs(next, trie, graph, queriesByNode, ans));
  trie.insert(curr, -1);
}

class Trie {
  constructor() {
    this.root = new TrieNode();
  }
  
  insert(num, d) {
    let node = this.root;
    
    for (let i = 17; i >= 0; i--) {
      const bit = (num >> i) & 1;
      if (!node.next[bit]) node.next[bit] = new TrieNode();
      
      node = node.next[bit];
      node.count += d;
    }
  }
  
  findMaxXOR(num) {
    let node = this.root;
    let max = 0;
    
    for (let i = 17; i >= 0; i--) {
      const bit = (num >> i) & 1;
      if (node.next[1 - bit] && node.next[1 - bit].count > 0) {
        max |= (1 << i);
        node = node.next[1 - bit];
      }
      
      else node = node.next[bit];
    }
    
    return max;
  }
}

class TrieNode {
  constructor() {
    this.next = [null, null];
    this.count = 0;
  }
}

File: 1938-maximum-genetic-difference-query/1938-maximum-genetic-difference-query.py
class Trie:
    def __init__(self):
        self.root = {}
        
    def insert(self, num, d):
        node = self.root
        for i in range(17, -1, -1):
            bit = (num >> i) & 1
            if bit not in node:
                node[bit] = {"cnt": 0}
            node = node[bit]
            node["cnt"] += d
            
    def find(self, num):
        node, ans = self.root, 0
        for i in range(17, -1, -1):
            bit = (num >> i) & 1
            if 1-bit in node and node[1-bit]["cnt"] > 0:
                ans |= (1 << i)
                node = node[1-bit]
            else:
                node = node[bit]
        return ans
      
    
class Solution:
    def maxGeneticDifference(self, parents: List[int], queries: List[List[int]]) -> List[int]:
        m, n, root = len(parents), len(queries), -1
        graph, queriesByNode = [[] for _ in range(m)], [[] for _ in range(m)]
        ans = [-1] * n
        trie = Trie()
        
        for i, p in enumerate(parents):
            if p == -1:
                root = i
            else:
                graph[p].append(i)
                
        for i, [node, val] in enumerate(queries):
            queriesByNode[node].append((val, i))
            
        def dfs(cur):
            trie.insert(cur, 1)
            for val, i in queriesByNode[cur]:
                ans[i] = trie.find(val)
            for nxt in graph[cur]:
                dfs(nxt)
            trie.insert(cur, -1)
            
        
        dfs(root)
        return ans